### DISCUSSION TITLE
How to remove unused .dlls files from packaged app?

### DISCUSSION THREAD

**Discussion by ahmed4end** [pyinstaller/pyinstaller#5910] [Help]
there are too many dlls in the dir of my pakage i tried to delete them one at a time and ran the app.exe and it actually worked with some of the dlls so the question here how to tell which ones are good to delete as they are not used so i have less files in the package dir ?

**Comment by bwoodsend** [comment]
Put:
```python
import psutil
this_process = psutil.Process()
used_dlls = [i.path for i in this_process.memory_maps()]
# Do something with this list of DLLs. e.g. Print them/write to file/...
[print(i) for i in used_dlls]
```
at the very end of your main Python file then rerun PyInstaller and rerun the app. This will show you all DLLs which are loaded into Python.

BUT BEWARE! DLLs can be conditionally loaded depending on hardware. And you may have 2<sup>nd</sup> copies of some of the more standard MSVC runtime libraries in `system32` which will be loaded instead, look like system libraries, and you will mistakenly delete the copies in your app because it appears not to need them. Then you're in a horrible situation to debug because your app will work perfectly on machines with those runtime DLLs (in particular your own) but fail to other machines. What's more, some of those DLLs are very common in standard developer tools like Git bash for Windows so the only people who can see the issue will be non-developers and likely not have a clue how to fix it or run your diagnostic tests for you. You'll find yourself teasing details out of someone whose diagnostics abilities extend to *try switching it off and on again* and has no idea what a DLL is - let alone how to debug them.

For this reason I recommend that you don't *just delete* anything if you can't see where it came from.

**Reply by ishipachev** [reply]
I would like to reiterate on this.

One specific example is OpenCV. Just reading and showing an image increases the size of the project. Significantly. From 20MB to 800MB. 

Having something like:
```python
from cv2 import imread, imwrite
```
still brings everything related to cv2 to the project. Even ffmpeg will be bundled with .exe as a .dll.

The question is:

1. Does pyinstaller analyzes which specific modules to include and if it reduce it to specific functions/modules but not the whole package?
2. If yes, than could it be that all OpenCV calls got through some single .dll which brings the rest and pyinstaller just could not track it down through .dll so it pulls all dependencies? 

I'm thinking what could be a possible solution which helps to establish CI integration process without diving deep into real-time analyzis as it was suggested above (good suggestion though!)? 
I know that delivering python bases solutions to customers not in the form of packages is a pain and should be avoided by any cost, but here we are, here I am :)

I'm aware that there is a way to decouple things by manually building exact functions of opencv somewhere between python and opencv. But again, any new addition will require an engineer to go into cpp code and rebuild an intermediate layer between python and opencv. Which again is not a good solution in terms of CI and supporting cost.

**Reply by rokm** [reply]
We analyze the module/package imports, so the granularity is at python module level (unless there are many indirect imports, in which case we need to resort to collecting everything). Which does not really help you with OpenCV, which has a single large binary extension (`cv2.cv2`), so it's pretty much all or nothing. And we also need to collect `opencv_videoio_ffmpeg` DLL, because we have no way of knowing if your application uses functions that require it or not.

That said, collected `cv2` directory should be around 100 MB, so any additional increases are probably coming from somewhere else; could be dependencies that `cv2` is pulling in, such as `numpy`, or some optional dependencies and their optional dependencies. But even so, a build of simple program containing just (`from cv2 import imread, imwrite`) comes out at 147 MB on my test machine.

**Reply by ishipachev** [reply]
@rokm 
Thank you for your comments.
I suspected that all OpenCV .dlls are connected through some intermediate binary.
And you are right, OpenCV pulls numpy, mkl, ffpmpeg and a lot of other stuff with it.

I believe it may be also connected with the way conda (in my case) install OpenCV precompiled binaries and its dependencies. Yes, numpy is included I believe.

I guess at the end for my case (using conda) will be: compile OpenCV with core functionality inside this conda environment and hope that this functionality will cover the most of it and have small chance to get back to this.

I will try to create a separate environment and experiment with various builds of OpenCV to minimize its binary footprint in the resulted package produced by pyinstaller.
